shell/browser/notifications/mac/notification_center_delegate.mm

```diff
@@ -6,63 +6,151 @@
+        const char* key = [_replyKey UTF8String];
+        stm << L"[";  // json array open brace
+        stm << L"{";  // open brace for key-value pair
+        stm << L"\"" << (key ? key : "") << L"\"";
+        stm << L":";  // json delimeter (i.e colon) between key value
+        stm << L"\"" << reply << L"\"";
+        stm << L"}";  // close brace for key-value pair
+        stm << L"]";  // json array close brace
+      }
+      app->Emit("notification-activation", [notif.identifier UTF8String],
+                action ? action : "", reply ? 1 : 0, reply ? stm.str() : "");
+    }
 
-  if (getenv("ELECTRON_DEBUG_NOTIFICATIONS")) {
-    LOG(INFO) << "Notification activated (" << [notif.identifier UTF8String]
-              << ")";
-  }
+    auto* notification = presenter_->GetNotification(notif);
 
-  if (notification) {
-    // Ref:
-    // https://developer.apple.com/documentation/foundation/nsusernotificationactivationtype?language=objc
-    if (notif.activationType ==
-        NSUserNotificationActivationTypeContentsClicked) {
-      notification->NotificationClicked();
-    } else if (notif.activationType ==
-               NSUserNotificationActivationTypeActionButtonClicked) {
-      notification->NotificationActivated();
-    } else if (notif.activationType ==
-               NSUserNotificationActivationTypeReplied) {
-      notification->NotificationReplied([notif.response.string UTF8String]);
-    } else {
+    if (getenv("ELECTRON_DEBUG_NOTIFICATIONS")) {
+      LOG(INFO) << "Notification activated (" << [notif.identifier UTF8String]
+                << ")";
+    }
+
+    if (notification) {
+      // Ref:
+      // https://developer.apple.com/documentation/foundation/nsusernotificationactivationtype?language=objc
       if (notif.activationType ==
-          NSUserNotificationActivationTypeAdditionalActionClicked) {
-        notification->NotificationActivated([notif additionalActivationAction]);
+          NSUserNotificationActivationTypeContentsClicked) {
+        notification->NotificationClicked();
+      } else if (notif.activationType ==
+                 NSUserNotificationActivationTypeActionButtonClicked) {
+        notification->NotificationActivated();
+      } else if (notif.activationType ==
+                 NSUserNotificationActivationTypeReplied) {
+        notification->NotificationReplied([notif.response.string UTF8String]);
+      } else {
+        if (notif.activationType ==
+            NSUserNotificationActivationTypeAdditionalActionClicked) {
+          notification->NotificationActivated(
+              [notif additionalActivationAction]);
+        }
       }
     }
-  }
+  });
+}
+
+- (void)userNotificationCenter:(NSUserNotificationCenter*)center
+        didDeliverNotification:(NSUserNotification*)notification {
+  [self didDeliverNotification:notification];
+  // Clear pending notification
+  presenter_->CloseNotificationWithId("![DELETE]");
```

Comment:
What is the significance of `![DELETE]` here?

Reply:
Just internal token which widely used in current implementation to mark notification ID which is not exist or not reachable. In current implementation it is used for potential clearing notification in **in_pending_deletion_** state. In it's turn this strange on the first state is used to implement support renotify property.